File: code/python/01-basic-llm-call/main.py
"""Token counting for OpenAI-compatible models using tiktoken.

Shows how to count tokens for a plain string and for a messages array
(the format used by chat.completions.create). The messages array count
mirrors what the API actually charges you for.
"""

from __future__ import annotations

import tiktoken

MODEL = "gpt-5.5"


def _get_encoding(model: str):
    """Return a tiktoken encoding for the model, with a safe fallback."""
    try:
        return tiktoken.encoding_for_model(model)
    except (KeyError, ValueError):
        # Some tiktoken versions may not know about newer model aliases.
        # Fall back to the cl100k_base encoding used by OpenAI chat models.
        return tiktoken.get_encoding("cl100k_base")


def count_tokens(text: str, model: str = MODEL) -> int:
    """Return the number of tokens in *text* for the given model."""
    enc = _get_encoding(model)
    return len(enc.encode(text))


def count_messages_tokens(messages: list[dict], model: str = MODEL) -> int:
    """Return the token cost of a messages array for chat completions.

    Accounts for the per-message overhead (3 tokens) and reply primer (3
    tokens) that the API adds automatically.
    See: https://platform.openai.com/docs/guides/chat/managing-tokens
    """
    enc = _get_encoding(model)
    tokens_per_message = 3
    tokens_per_name = 1
    total = 0
    for message in messages:
        total += tokens_per_message
        for key, value in message.items():
            total += len(enc.encode(value))
            if key == "name":
                total += tokens_per_name
    total += 3  # reply is primed with <|start|>assistant<|message|>
    return total


def main() -> None:
    text = "The quick brown fox jumps over the lazy dog."
    print(f"Text  : {text!r}")
    print(f"Tokens: {count_tokens(text)}")

    messages = [
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "What is the capital of France?"},
    ]
    print(f"\nMessages array token count: {count_messages_tokens(messages)}")


if __name__ == "__main__":
    main()

File: code/go/01-basic-llm-call/main.go
// Token counting for OpenAI-compatible models using tiktoken-go.
//
// Shows how to count tokens for a plain string and for a messages slice
// (the format used by chat.completions.create). The messages slice count
// mirrors what the API actually charges you for.
package main

import (
	"fmt"
	"log"

	tiktoken "github.com/pkoukk/tiktoken-go"
)

const model = "gpt-4o"

// countTokens returns the number of tokens in text for the given model.
func countTokens(text, model string) (int, error) {
	enc, err := tiktoken.EncodingForModel(model)
	if err != nil {
		return 0, err
	}
	return len(enc.Encode(text, nil, nil)), nil
}

// message mirrors the OpenAI chat completions messages format.
type message struct {
	Role    string
	Name    string // optional
	Content string
}

// countMessagesTokens returns the token cost of a messages slice for chat
// completions. Accounts for the per-message overhead (3 tokens) and reply
// primer (3 tokens) that the API adds automatically.
// See: https://platform.openai.com/docs/guides/chat/managing-tokens
func countMessagesTokens(messages []message, model string) (int, error) {
	enc, err := tiktoken.EncodingForModel(model)
	if err != nil {
		return 0, err
	}
	const tokensPerMessage = 3
	const tokensPerName = 1
	total := 0
	for _, msg := range messages {
		total += tokensPerMessage
		total += len(enc.Encode(msg.Role, nil, nil))
		total += len(enc.Encode(msg.Content, nil, nil))
		if msg.Name != "" {
			total += len(enc.Encode(msg.Name, nil, nil))
			total += tokensPerName
		}
	}
	total += 3 // reply is primed with <|start|>assistant<|message|>
	return total, nil
}

func main() {
	text := "The quick brown fox jumps over the lazy dog."
	n, err := countTokens(text, model)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("Text  : %q\n", text)
	fmt.Printf("Tokens: %d\n", n)

	messages := []message{
		{Role: "system", Content: "You are a helpful assistant."},
		{Role: "user", Content: "What is the capital of France?"},
	}
	total, err := countMessagesTokens(messages, model)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("\nMessages array token count: %d\n", total)
}
